### PR TITLE
Implement STARTTLS event; Fix #113

### LINKED ISSUE #113: Implement STARTTLS for circuits.net.sockets
Maybe we can leverage this [Tornado STARTTLS example](https://gist.github.com/weaver/293449).

I'd like to see an API such as this:

``` #!python
"""Simple TLS Echo Server"""


from circuits import Component
from circuits.net.tls import starttls, TLS
from circuits.net.sockets import TCPServer


class TLSEchoServer(Component):

    def init(self, bind):
        self.transport = (TCPServer(*bind) + TLS()).register(self)

    def read(self, sock, data):
        if data.strip.upper() == "STARTTLS":
            return self.fire(starttls())
        return data


TLSEchoServer(("0.0.0.0", 8000)).run()
```

Thoughts?

### PR DESCRIPTION
**Pull request by spaceone**
```
from circuits import Component, Debugger
from circuits.net.sockets import TCPServer, write, starttls

class TLSEchoServer(Component):

    def init(self, bind):
        self.transport = TCPServer(bind,
            certfile="/circuits/tests/net/cert.pem",
        ).register(self)

    def connect(self, sock, foo, bar):
        self.fire(write(sock, b'* OK [CAPABILITY STARTTLS] Dovecot ready.\r\n'))

    def read(self, sock, data):
        if data.strip().upper() == b". STARTTLS":
            yield self.fire(write(sock, b'. OK Begin TLS negotiation now.'))
            self.fire(starttls(sock))
            return
        if data.strip() == b'. CAPABILITY':
            self.fire(write(sock, b'* CAPABILITY STARTTLS \r\n. OK Pre-login capabilities listed, post-login capabilities have more.\r\n'))
        else:
            self.fire(write(sock, b'Thanks for TLS encrypting! You wrote: %r' % (data,)))

(TLSEchoServer(("0.0.0.0", 8000)) + Debugger()).run()
```
$ openssl s_client -connect localhost:8000 -starttls imap

### CODE REVIEW

File: circuits/net/sockets.py [hunk 1]
```diff
@@ -607,21 +594,47 @@ def on_error(sock, err):
...
+        )
+
+        for _ in do_handshake(sslsock, self._on_accept_done, self._on_handshake_error):
+            yield _
+
+    def _on_accept_done(self, sock):
+        sock.setblocking(False)
+        self._poller.addReader(self, sock)
+        self._clients.append(sock)
+        self.fire(connect(sock, *sock.getpeername()))
```
Comment (spaceone) [inline circuits/net/sockets.py]: Add check for already started starttls and don't fire connect again.

File: circuits/net/sockets.py [hunk 2]
```diff
@@ -607,21 +594,47 @@ def on_error(sock, err):
...
+
+    def _on_handshake_error(self, sock, err):
+        self.fire(error(sock, err))
+        self._close(sock)
+
+    @handler('starttls')
+    def starttls(self, sock):
+        if not HAS_SSL:
+            return
+        if self.__starttls:
```
Comment (spaceone) [inline circuits/net/sockets.py]: Oh, this seems wrong. Should be a dict of sock -> bool.

**Comment by prologic** [comment]
Nice!

File: circuits/net/sockets.py [hunk 3]
```diff
@@ -607,21 +597,47 @@ def on_error(sock, err):
...
+        self.fire(connect(sock, *sock.getpeername()))
+
+    def _on_handshake_error(self, sock, err):
+        self.fire(error(sock, err))
+        self._close(sock)
+
+    @handler('starttls')
+    def starttls(self, sock):
+        if not HAS_SSL:
+            return  # TODO: emit a warning?
```
Comment (prologic) [inline circuits/net/sockets.py]: Just raise an appropriate exception maybe?

Comment (spaceone) [inline circuits/net/sockets.py]: @prologic Done that.

**Pull request merged**